### ISSUE TITLE
bug: occasional failures freeswitch binding to rtp port causing call failure

### ISSUE THREAD

**Issue by davehorton** [jambonz/jambonz-feature-server#202]
During load testing, intermittent call failures can be traced to freeswitch returning a 488

```
2022-12-29 19:48:04.686880 send 693 bytes to udp/[10.0.135.63]:5080 at 19:48:04.686814:
INVITE sip:drachtio@10.0.135.63:5080 SIP/2.0^M
Via: SIP/2.0/UDP 10.0.135.63:5070;rport;branch=z9hG4bKaKrmgB7SH1QjH^M
Max-Forwards: 70^M
From: <sip:10.0.135.63:5070>;tag=XZy8Sctvm7mHj^M
To: <sip:drachtio@10.0.135.63:5080>^M
Call-ID: 8d22bd16-0254-123c-9aae-0632bc4a7412^M
CSeq: 61617850 INVITE^M
Contact: <sip:10.0.135.63:5070>^M
User-Agent: drachtio-fsmrf:9c027e1e-dbab-493b-9545-47bc60f5e55d^M
Content-Type: application/sdp^M
Content-Length: 160^M
X-Jambones-Call-ID: 8c184f37-0254-123c-5fb1-0632bc4a7412^M
X-esl-outbound: 10.0.135.63:37447^M
^M
v=0^M
o=user1 53655765 2353687637 IN IP4 10.0.135.63^M
s=-^M
c=IN IP4 10.0.135.63^M
t=0 0^M
m=audio 52012 RTP/AVP 0^M
a=rtpmap:0 PCMU/8000^M
a=sendrecv^M
a=rtcp:52013

2022-12-29 19:48:04.688957 recv 689 bytes from udp/[10.0.135.63]:5080 at 19:48:04.688896:
SIP/2.0 488 Not Acceptable Here^M
Via: SIP/2.0/UDP 10.0.135.63:5070;rport=5070;branch=z9hG4bKaKrmgB7SH1QjH^M
Max-Forwards: 70^M
From: <sip:10.0.135.63:5070>;tag=XZy8Sctvm7mHj^M
To: <sip:drachtio@10.0.135.63:5080>;tag=BFvKXNmtU25Qa^M
Call-ID: 8d22bd16-0254-123c-9aae-0632bc4a7412^M
CSeq: 61617850 INVITE^M
User-Agent: Jambonz media gateway^M
Accept: application/sdp^M
Allow: INVITE, ACK, BYE, CANCEL, OPTIONS, MESSAGE, INFO, UPDATE, REGISTER, REFER, NOTIFY^M
Supported: path, replaces^M
Allow-Events: talk, hold, conference, refer^M
Reason: Q.850;cause=88;text="INCOMPATIBLE_DESTINATION"^M
Content-Length: 0^M
Remote-Party-ID: "drachtio" <sip:drachtio@10.0.135.63>;party=calling;privacy=off;screen=no^M
```

If we look into this on the freeswitch log, we see that it failed to bind to an rtp port, or an rtcp port.  In the example above, it was an rtp port

```
2022-12-29 19:48:04.670474 [ERR] switch_core_media.c:9669 AUDIO RTP REPORTS ERROR: [Bind Error! 10.0.135.63:37814]
```

Now what is listening on that port?  It turns out it is feature-server:

```
admin@ip-10-0-135-63:~/apps/sbc-inbound$ lsof -i udp:37814
COMMAND     PID  USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
node\x20/ 21654 admin   31u  IPv4 421279      0t0  UDP *:37814
admin@ip-10-0-135-63:~/apps/sbc-inbound$ ps -elf | grep 21654
0 S admin      21654    4671  4  80   0 - 2866947 -    19:46 ?        00:00:29 node /home/admin/apps/jambonz-feature-server/app.js
```

Is feature-server listening on any other port?  No, just that one:
```
$ lsof -i -P -a -p 21654 | grep UDP
node\x20/ 21654 admin   31u  IPv4 421279      0t0  UDP *:37814
```

What about sbc-inbound?  Turns out it is listening on a udp port as well:
```
$ lsof -i -P -a -p 21590 | grep UDP
node\x20/ 21590 admin   24u  IPv4 419375      0t0  UDP *:37251
```

OK, what about sbc-outbound? Yup, also listening
```
$ ps -elf |grep sbc-outbound
0 S admin      21583    4671  0  80   0 - 2792280 -    19:46 ?        00:00:02 node /home/admin/apps/sbc-outbound/app.js
0 S admin      29759    5517  0  80   0 -  1318 -      20:05 pts/0    00:00:00 grep sbc-outbound
admin@ip-10-0-135-63:~/apps/sbc-inbound$ lsof -i -P -a -p 21583 | grep UDP
node\x20/ 21583 admin   25u  IPv4 424184      0t0  UDP *:57728
```

sbc-sip-sidecar is _not_ listening on a udp port, but sbc-rtpengine-sidecar is:
```
ps -elf |grep sbc-rtpengine-sidecar
0 S admin      21609    4671  0  80   0 - 159242 -     19:46 ?        00:00:01 node /home/admin/apps/sbc-rtpengine-sidecar/app.js
0 S admin      29804    5517  0  80   0 -  1319 -      20:06 pts/0    00:00:00 grep sbc-rtpengine-sidecar
admin@ip-10-0-135-63:~/apps/sbc-inbound$ lsof -i -P -a -p 21609 | grep UDP
node\x20/ 21609 admin   20u  IPv4 427035      0t0  UDP *:22223
```

jambonz-api-server and sbc-call-router are _not_ listening on any udp ports.

I suspect the problem on sbc-inbound and sbc-outbound may be rtpengine-client, which binds to a random UDP port, unless a specific port is specified.  We should pass a specific local port that is outside the range that freeswitch is configured to use (25000-39000), and also outside the range that rtpengine is configured to use (40000-60000).

This doesn't explain sbc-rtpengine-sidecar or feature-server though, because those do not use that library.

**Comment by davehorton** [comment]
initial testing validates that for sbc-inbound and sbc-outbound the culprit is rtpengine-utils.  Theoretically we can fix this by passing a local port to bind to that is out of the range of freeswitch and rtpengine, but in the case of multiple sbc-inbound instances there would still be a collision..

**Comment by davehorton** [comment]
the [portfinder][https://www.npmjs.com/package/portfinder] package might be useful, though I could still see collisions happening..

**Comment by davehorton** [comment]
something else interesting..when I first start up feature-server it is not binding to a port.  Only once it starts handling calls does it bind to a udp port

**Comment by davehorton** [comment]
Another part of the mystery solved: sbc-rtpengine-sidecar binds to a udp port to receive dtmf notifications from rtpengine.  However, it always binds to port 22223 which is safely out of the problem range.  So no issue there.

**Comment by davehorton** [comment]
ok, in pm2.io we have the `process.env.NODE_APP_INSTANCE` var that we can use.  So if for instance there were 4 instances of sbc-inbound running they would each have unique values for that env of 0 through 3.  So we can then have an optional env which is a base port number (safely outside the problem range) to which we add that env value, if present, and then use that resulting number as the local port to bind to.  That should solve the problem for sbc-inbound and sbc-outbound.  Only remaining mystery is why/where feature-server is binding to a udp port..

**Comment by davehorton** [comment]
fixes have been checked in for @jambonz/rtpengine-utils and to sbc-inbound and sbc-outbound.  The mystery about jambonz-feature-server has also been solved -- the issue was that the oltp tracing client by default sends traces to jaeger using UDP, and this was grabbing the random port.  This can be changed to http transport, and this change has been successfully tested.

These fixes shall appear in v0.7.8-3
